packages/core/src/fund.ts

```diff
@@ -74,31 +73,41 @@ export const estimateExecutionGas = async (
     actions.filter((action) => isSetStorageAction(action)).length
   )
 
-  const deployedProxyPromises = actions
-    .filter((action) => isDeployContractAction(action))
-    .map(async (action) => {
-      return (await isContractDeployed(
-        getDefaultProxyAddress(
-          claimer,
-          organizationID,
-          projectName,
-          action.referenceName
-        ),
-        provider
-      ))
+  const managerAddress = getChugSplashManagerAddress(
+    parsedConfig.options.claimer,
+    parsedConfig.options.organizationID
+  )
+
+  const deployedProxyPromises = Object.values(parsedConfig.contracts).map(
+    async (contract) =>
+      // If the proxy has already been deployed, then estimate 0 gas. Otherwise, estimate 550k for the default proxy.
+      (await isContractDeployed(contract.proxy, provider))
         ? ethers.BigNumber.from(0)
         : ethers.BigNumber.from(550_000)
-    })
+  )
 
   const deployedContractPromises = actions
     .filter((action) => isDeployContractAction(action))
-    .map(async (action: DeployContractAction) =>
-      ethers.BigNumber.from(350_000).add(
-        await provider.estimateGas({
-          data: action.code,
-        })
+    .map(async (action: DeployContractAction) => {
+      const implementationAddress = utils.getCreate2Address(
+        managerAddress,
+        ethers.constants.HashZero,
+        utils.solidityKeccak256(['bytes'], [action.code])
       )
-    )
+
+      // If the implementation has already been deployed, then estimate 0 gas. Otherwise, estimate the gas to deploy the implementation.
+      const deployCost = (await isContractDeployed(
+        implementationAddress,
+        provider
+      ))
+        ? ethers.BigNumber.from(0)
+        : await provider.estimateGas({
+            data: action.code,
+          })
+      // We also tack on an extra 100k gas for each proxy target (including any that are not being upgraded) to account
+      // for the variable cost of the `initiateBundleExecution` and `completeBundleExecution` functions.
+      return deployCost.add(ethers.BigNumber.from(100_000))
```

Comment:
Accounts for the variable cost of the initiate and complete functions.

Reply:
I think this is a good idea, but I think we should put it somewhere outside of this `map` statement. This is because the `actionBundle`, which is being iterated over here, can have a different number of proxies than the `targetBundle`. (See [here](https://github.com/chugsplash/chugsplash/pull/617) for an explanation).

I'd probably just do something like this:
```
const initiateAndCompleteCost = ethers.BigNumber.from(100_000).mul(bundles.targetBundle.targets.length)
```